Fix issue where updating a single unit results in other units being turned off

blakerouse opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

Fixes an issue in the `managerV2` where it would not call reload with all inputs, resulting in inputs being turned off when they should stay running.

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

Fixes an issue where if one input changes it turns off all the other inputs.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Closes https://github.com/elastic/elastic-agent/issues/2232

elasticmachine commented:
Pinging @elastic/elastic-agent (Team:Elastic-Agent)

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > There is a new build on-going so the previous on-going builds have been aborted.
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34504/detail/PR-34504/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34504/detail/PR-34504/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34504/detail/PR-34504/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34504/detail/PR-34504/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2023-02-07T15:49:50.302Z&rangeTo=2023-02-07T16:09:50.302Z&transactionName=BUILD+Beats%2Fbeats%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=188ddaedab4e748780113845e5aabdbf&transactionId=f2ea626445c81052)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-02-07T15:59:50.302+0000


* Duration: 11 min 41 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 3  |
| Skipped      | 0 |
| Total        | 3   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->
















## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

leehinman reviewed (approved):
LGTM

cmacknz commented on CHANGELOG.next.asciidoc:
```diff
@@ -53,12 +53,13 @@ https://github.com/elastic/beats/compare/v8.2.0\...main[Check the HEAD diff]
 - Fix concurrent map writes when system/process code called from reporter code {pull}32491[32491]
 - Log errors from the Elastic Agent V2 client errors channel. Avoids blocking when error occurs communicating with the Elastic Agent. {pull}34392[34392]
 - Only log publish event messages in trace log level under elastic-agent. {pull}34391[34391]
+- Fix issue where updating a single unit results in other units being turned off. {pull}34504[34504]
```
```suggestion
- Fix issue where updating a single Elastic Agent configuration unit results in other units being turned off. {pull}34504[34504]
```

Let's directly mention agent directly since this is the Beats repository.

cmacknz reviewed (commented):
Needs a test case (and to fix the existing unit tests it seems). You will have to manually backport to the 8.7.0 branch if this doesn't merge by the end of today.

blakerouse commented:
@cmacknz The unit test is actually testing that path, that is why the test is failing. The code is actually not working for stopping, just noticed that in local testing as well. Working on a fix and will do some more testing.

blakerouse commented:
@cmacknz Local testing is looking good, would like to get an extra set of testing on this PR before merging. Could you give it a test as well?

cmacknz commented:
> The unit test is actually testing that path, that is why the test is failing. The code is actually not working for stopping, just noticed that in local testing as well. Working on a fix and will do some more testing.

> @cmacknz Local testing is looking good, would like to get an extra set of testing on this PR before merging. Could you give it a test as well?

I don't see any new tests that would have caught the problem we are fixing here. I am concerned this bug might come back because there is nothing stopping that from happening. Can we write a test that ensures that a change that results in no unit config changes (like a log level change) preserves the set of input units that should have been running?

This needs to make the 8.6.2 feature freeze which is EOD Thursday. The 8.7 FF is today but since this is a bug fix we can backport it in after that. We have time to add an additional test I think.

blakerouse commented:
@cmacknz I modified the current test to catch the bad case. I verified this by removing all the changes I had added in managerV2 and ensuring that the test failed. Then I placed my modified code back into the managerV2 and the test past, so catching this error is now covered by the unit test.

cmacknz commented:
Thanks! Test LGTM.
